common/src/main/java/org/terracotta/angela/common/tcconfig/holders/TcConfigHolder.java

```diff
@@ -106,76 +107,75 @@ public void writeTcConfigFile(File kitDir, final String tcConfigFilename) {
-          int tsaPort = tsaPortNode == null ? defaultTsaPort() : Integer.parseInt(tsaPortNode.getTextContent());
+        Node hostNode = (Node) xPath.evaluate("@host", server, XPathConstants.NODE);
+        String hostname =
+            hostNode == null || hostNode.getTextContent().equals("%i") || hostNode.getTextContent()
+                .equals("%h") ? "localhost" : hostNode.getTextContent();
 
-          Node jmxPortNode = (Node) xPath.evaluate("*[name()='jmx-port']", server, XPathConstants.NODE);
-          int jmxPort = jmxPortNode == null ? defaultJmxPort() : Integer.parseInt(jmxPortNode.getTextContent());
+        Node nameNode = (Node) xPath.evaluate("@name", server, XPathConstants.NODE);
 
-          Node tsaGroupPortNode = (Node) xPath.evaluate("*[name()='tsa-group-port']", server, XPathConstants.NODE);
-          int tsaGroupPort = tsaGroupPortNode == null ? defaultTsaGroupPort() : Integer.parseInt(tsaGroupPortNode.getTextContent());
+        //TODO : create client and send command to get free port -> can't connect ? agent exception!
+        // add into xml the port!!
+        // log it!!!
+        // remove below updatePorts method
 
-          Node managementPortNode = (Node) xPath.evaluate("*[name()='management-port']", server, XPathConstants.NODE);
-          int managementPort = managementPortNode == null ? defaultManagementPort() : Integer.parseInt(managementPortNode.getTextContent());
+        int tsaPort = getOrAllocate("tsa-port", server, i, xPath, portAllocator);
+        int jmxPort = support.apply("jmx-port") ? getOrAllocate("jmx-port", server, i, xPath, portAllocator) : 0;
+        int tsaGroupPort = getOrAllocate("tsa-group-port", server, i, xPath, portAllocator);
+        int managementPort = support.apply("management-port") ? getOrAllocate("management-port", server, i, xPath, portAllocator) : 0;
 
-          String symbolicName = nameNode == null ? hostname + ":" + tsaPort : nameNode.getTextContent();
+        String symbolicName = nameNode == null ? hostname + ":" + tsaPort : nameNode.getTextContent();
 
-          TerracottaServer terracottaServer = TerracottaServer
-              .server(symbolicName, hostname)
-              .tsaPort(tsaPort)
-              .tsaGroupPort(tsaGroupPort)
-              .managementPort(managementPort)
-              .jmxPort(jmxPort);
-          servers.add(terracottaServer);
-        }
-      } catch (Exception e) {
-        throw new RuntimeException("Cannot parse tc-config xml", e);
+        TerracottaServer terracottaServer = TerracottaServer
+            .server(symbolicName, hostname)
+            .tsaPort(tsaPort)
+            .tsaGroupPort(tsaGroupPort)
+            .managementPort(managementPort)
+            .jmxPort(jmxPort);
+        servers.add(terracottaServer);
       }
-      this.servers.addAll(servers);
+    } catch (Exception e) {
+      throw new RuntimeException("Cannot parse tc-config xml", e);
     }
-    return servers;
-  }
```

Comment:
Fixed the code above that was not correctly supporting the OSS/EE diffs plus the rewriting of the xml content